Remove dependency of Roberta in Blenderbot

rchan26 opened the pull request:
Hi @sgugger,

This PR looks to address https://github.com/huggingface/transformers/issues/19303: the RobertaTokenizer dependency has been removed from `BlenderbotTokenizer` and the RobertaFastTokenizer dependency has been removed from `BlenderbotTokenizerFast`.

I did encounter an error when running `pytest tests/models/blenderbot/test_tokenization_blenderbot.py`, and I got the following error:
```
========================================================================= test session starts =========================================================================
platform darwin -- Python 3.10.4, pytest-7.1.3, pluggy-1.0.0
rootdir: /Users/rchan/Library/CloudStorage/OneDrive-TheAlanTuringInstitute/huggingface/transformers, configfile: setup.cfg
collected 4 items

tests/models/blenderbot/test_tokenization_blenderbot.py F...                                                                                                    [100%]

============================================================================== FAILURES ===============================================================================
___________________________________________________ Blenderbot3BTokenizerTests.test_3B_tokenization_same_as_parlai ____________________________________________________

self = <tests.models.blenderbot.test_tokenization_blenderbot.Blenderbot3BTokenizerTests testMethod=test_3B_tokenization_same_as_parlai>

    def test_3B_tokenization_same_as_parlai(self):
        assert self.tokenizer_3b.add_prefix_space
>       assert self.tokenizer_3b([" Sam", "Sam"]).input_ids == [[5502, 2], [5502, 2]]
E       assert [[1, 5502, 2], [1, 5502, 2]] == [[5502, 2], [5502, 2]]
E         At index 0 diff: [1, 5502, 2] != [5502, 2]
E         Use -v to get more diff

tests/models/blenderbot/test_tokenization_blenderbot.py:48: AssertionError
------------------------------------------------------------------------ Captured stderr call -------------------------------------------------------------------------
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
========================================================================== warnings summary ===========================================================================
src/transformers/testing_utils.py:28
  /Users/rchan/Library/CloudStorage/OneDrive-TheAlanTuringInstitute/huggingface/transformers/src/transformers/testing_utils.py:28: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.util import strtobool

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================= short test summary info =======================================================================
FAILED tests/models/blenderbot/test_tokenization_blenderbot.py::Blenderbot3BTokenizerTests::test_3B_tokenization_same_as_parlai - assert [[1, 5502, 2], [1, 5502, 2]...
=============================================================== 1 failed, 3 passed, 1 warning in 2.15s ================================================================
```

Any idea on what I have done wrong here?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

rchan26 committed:
Remove dependency of Roberta in Blenderbot

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19411). All of your documentation changes will be reflected on that endpoint.

sgugger commented on src/transformers/models/blenderbot/tokenization_blenderbot.py:
```diff
@@ -14,10 +14,18 @@
-from ..roberta.tokenization_roberta import RobertaTokenizer
+
+
+# from ..roberta.tokenization_roberta import RobertaTokenizer
```
Should be completely removed :-)

sgugger commented on src/transformers/models/blenderbot/tokenization_blenderbot.py:
```diff
@@ -43,53 +51,365 @@
+    return pairs
+
+
+# Copied from transformers.models.roberta.tokenization_roberta.RobertaTokenizer with roberta-base->facebook/blenderbot-3B, Roberta->Blenderbot, RoBERTa->Blenderbot
```
In this instance, you can't add a Copied from on the whole class as it deletes the two methods this specific subclass implements (`build_inputs_with_special_tokens` and `_build_conversation_input_ids`). So you need to put the Copied from on each method that is the same as in Roberta instead.

rchan26 replied:
Okay, I'll add '#Copied from...` for each of the methods :)

One question on copying over the methods from `RobertaTokenizer()`, I noticed that `RobertaTokenizer()` and `BlenderbotTokenizer()` both have a `build_inputs_with_special_tokens()` method. Currently, in the `RobertaTokenizer` class, we have the method:

```
def build_inputs_with_special_tokens(
        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
    ) -> List[int]:
        """
        Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and
        adding special tokens. A RoBERTa sequence has the following format:
        - single sequence: `<s> X </s>`
        - pair of sequences: `<s> A </s></s> B </s>`
        Args:
            token_ids_0 (`List[int]`):
                List of IDs to which the special tokens will be added.
            token_ids_1 (`List[int]`, *optional*):
                Optional second list of IDs for sequence pairs.
        Returns:
            `List[int]`: List of [input IDs](../glossary#input-ids) with the appropriate special tokens.
        """
        if token_ids_1 is None:
            return [self.cls_token_id] + token_ids_0 + [self.sep_token_id]
        cls = [self.cls_token_id]
        sep = [self.sep_token_id]
        return cls + token_ids_0 + sep + sep + token_ids_1 + sep
```
and in the `BlenderbotTokenizer` class, we have a method with the same name:
```
    def build_inputs_with_special_tokens(self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None):
        """
        Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and
        adding special tokens. A Blenderbot sequence has the following format:
        - single sequence: ` X </s>`
        Args:
            token_ids_0 (`List[int]`):
                List of IDs to which the special tokens will be added
            token_ids_1 (`List[int]`, *optional*):
                Will be ignored
        Returns:
            `List[int]`: list of [input IDs](../glossary#input-ids) with the appropriate special tokens.
        """
        return token_ids_0 + [self.eos_token_id]
```

In the case where there are conflicting method names, which one should I choose? I think originally in my PR before, I took the one from `RobertaTokenizer` and just copied this over.

sgugger commented on src/transformers/models/blenderbot/tokenization_blenderbot_fast.py:
```diff
@@ -44,51 +47,236 @@
-class BlenderbotTokenizerFast(RobertaTokenizerFast):
-    r"""
-    Construct a "fast" Blenderbot tokenizer (backed by HuggingFace's *tokenizers* library).
+# Copied from transformers.models.roberta.tokenization_roberta_fast.RobertaTokenizerFast with roberta-base->facebook/blenderbot-3B, Roberta->Blenderbot, RoBERTa->Blenderbot
```
Same here

sgugger reviewed (commented):
In this case, we can't have a global Copied from on the whole class as there are methods implemented by the subclass that are different from the parent class. So you need to put the Copied from on each method that is the same as in Roberta instead.

rchan26 committed:
Move Copied from statements to each method of the Roberta classes

rchan26 commented:
Hi @sgugger, I have now removed the global ` # Copied from` statements, and replaced them with Copied from statements to the individual methods that I am copying into the Blenderbot classes. This has resolved my earlier problem and now `pytest tests/models/blenderbot/test_tokenization_blenderbot.py` runs without error.

Currently, my PR now fails `python utils/check_copies.py` as there are two methods named `mask_token` in the `RobertaTokenizerFast`. This means that I currently have two `# Copied from transformers.models.roberta.tokenization_roberta_fast.RobertaTokenizerFast.mask_token` statements and so there's a matching problem.

How should I deal with the case where there are two methods with the same name?

rchan26 committed:
Remove copied from line for mask_token.setter

rchan26 committed:
update output from example in docs

rchan26 commented:
I've just seen that someone had a similar issue with copying over a method with a setter: https://github.com/huggingface/transformers/pull/19408#pullrequestreview-1134731877.

I have now followed the advice on this PR and have removed my Copied from statement on the setter for `mask_token`. Seems like all tests pass now 🙂

sgugger reviewed (approved):
Thanks a lot, looking great indeed!
